### PR TITLE
Form phone field validation as internationally-standardized format

### PR DESCRIPTION
**Pull request by kuzmany**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/1614
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

This bugfix/feature has been sponsored by @Webmecanik

[//]: # ( Required: )
#### Description:
Based on this feature request https://github.com/mautic/mautic/issues/1614 
This PR added validation based on https://github.com/giggsey/libphonenumber-for-php, already used in SMSBundle. 
Twillio require E.164 number formatting   https://support.twilio.com/hc/en-us/articles/223183008-Formatting-International-Phone-Numbers

#### Steps to test this PR:
1.  Create form
2.  Create Phone field. See properties tab and  option:  Validate as internationally-standardized format E.164. Enable it and also enable  Validation as  Required
image
3. Use form on your site.
4. See if form is submitted just after paste corect internation aformat (+421911111111 for example for Slovakia)

### CODE REVIEW

**Comment by npracht** [comment]
@justinfortes could we add that as candidate for 2.13.0 ?
Thanks !

File: app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        if (false === $phoneUtil->isValidNumber($phoneNumber)) {
+            $this->addViolation($value, $constraint);
+
+            return;
+        }
+
+        switch ($constraint->getType()) {
+            case PhoneNumber::FIXED_LINE:
```
Comment (Dreiser) [inline app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php]: This causes error. In use statement is "use libphonenumber\PhoneNumber as PhoneNumberObject;" so it needs to be used with PhoneNumberObject alias

**Comment by Dreiser** [comment]
@kuzmany I'm getting error 

> [2018-03-23 13:42:52] mautic.ERROR: PHP Error: Class 'Mautic\FormBundle\Validator\Constraint\PhoneNumber' not found - in file /var/www/html/app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php - at line 65 [] []

**Comment by kuzmany** [comment]
Please re-test
Label: Pending Feedback

**Comment by kuzmany** [comment]
Label: Ready to test

**Comment by javjim-mautic** [comment]
I am getting a error when pasting example number (+4219111111111) and using a known validnumber

"The phone number is not in internationally-standardized E.164 format. E.164 numbers can have a maximum of fifteen digits and are usually written as follows: [+][country code][subscriber number including area code]."

**Comment by kuzmany** [comment]
Label: WIP

**Comment by kuzmany** [comment]
@javjim-mautic I think you number is wrong
+421911 111 1111 translate to our slovak mobile number 0911 111 1111 but we allow just 0911 111 111.

**Comment by kuzmany** [comment]
@javjim-mautic  It was my mistake I wrote wrong phone number to description. 
I updated description please add to next milestone and re-test.
I checked updated number in https://www.twilio.com/console/lookup 
Thanks

**Comment by npracht** [comment]
@Dreiser & @alanhartless & @escopecz this is a **great improvement** to avoid issues on SMS integration/features use and manage clean data.
What is the status with it?

Can't we milestone it for 2.14 ?

**Comment by npracht** [comment]
@Dreiser can you review your review please? :)

**Comment by kuzmany** [comment]
@Araoo Fixed. Please test again. I find issue with PHP and bad condition with number type

**Comment by Araoo** [comment]
@kuzmany it's working for french numbers! I tried also Spanish and Belgium - it's working as well.

File: app/bundles/FormBundle/Form/Type/FormFieldTelType.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            [
+                'label'      => 'mautic.form.field.type.tel.international',
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
```
[the remark this answers is not in the document]
Comment (kuzmany) [reply app/bundles/FormBundle/Form/Type/FormFieldTelType.php]: That means we cannot use alias in config.php? https://github.com/mautic/mautic/pull/5400/files#diff-49a5edd5b79a5cd0fd0292ade1020ac2R271

File: app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_scalar($value) && !(is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedTypeException($value, 'string');
```
[the remark this answers is not in the document]
Comment (kuzmany) [reply app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php]: I am  not sure but use path to class in use: https://github.com/mautic/mautic/pull/5400/files/bb312dc67dd57bac520ab6f4d09d1f0b5ba30a71#diff-34498c36062aaacf746875fbeaeeeb3cR21

File: app/bundles/FormBundle/Form/Type/FormFieldTelType.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/FormBundle/Form/Type/FormFieldTelType.php]: We shouldn't use form aliases at all. Use full class name instead. https://github.com/symfony/symfony-docs/issues/5588

[the remark this answers is not in the document]
Comment (kuzmany) [reply app/bundles/FormBundle/Form/Type/FormFieldTelType.php]: That means we can remove also all config code for  field type? https://github.com/mautic/mautic/pull/5400/files#diff-49a5edd5b79a5cd0fd0292ade1020ac2R269

[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/FormBundle/Form/Type/FormFieldTelType.php]: I think we can remove the alias from there. The class will probably have to be present in the config. But give it a try and we'll know for sure :)

File: app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (escopecz) [reply app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php]: I missed that. Perfect!

File: app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            $validTypes = [];
+        }
+
+        if (count($validTypes)) {
+            $type = $phoneUtil->getNumberType($phoneNumber);
+
+            if (false === in_array($type, $validTypes)) {
+                $this->addViolation($value, $constraint);
+
+                return;
```
[the remark this answers is not in the document]
Comment (kuzmany) [reply app/bundles/FormBundle/Validator/Constraint/PhoneNumberConstraintValidator.php]: I optimize constraint classes just for phone number validation. Classes had taken from https://github.com/giggsey/libphonenumber-for-php and modify

File: app/bundles/FormBundle/Helper/FormFieldHelper.php [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <?php
 
 /*
- * @copyright   2014 Mautic Contributors. All rights reserved
+ * @copyright   2018 Mautic Contributors. All rights reserved
```
Comment (escopecz) [inline app/bundles/FormBundle/Helper/FormFieldHelper.php]: Actually, the copyrights dates should be set to the current date only for the newly created files. Not for the old ones. The year should state when the file was created and so when the copyright started.

Comment (kuzmany) [reply app/bundles/FormBundle/Helper/FormFieldHelper.php]: I know I just make mistake ;)Revert

**Pull request merged**